### PR TITLE
Fix incorrect nullable columns reporting in MySQL binary protocol

### LINKED ISSUE #56180: Tableau converts null in nullable numbers and dates to "0" or "01/01/1970" when using the MySQL Clickhouse connector
When using Tableau, and connecting to Clickhouse through the MySQL connector, we noticed that it doesn't handle `Nullable` fields very well.

When running this SQL in Tableau:

```sql
SELECT NULL AS number_nullable, NULL AS string_nullable, NULL AS date_nullable
UNION ALL
SELECT 123 AS number_nullable, 'abc' AS string_nullable, CURDATE() AS date_nullable;
```

To a regular MySQL instance (all good here!):

<p align="center"></p>

When running it to the MySQL Clickhouse instance:
- the `NULL` value in `date_nullable` becomes `01/01/1970`
- `NULL` value in `number_nullable` becomes `0`

<p align="center"> </p>

I am not able to reproduce this using other MySQL clients, only Tableau. When connecting with `mysql` in my terminal:

To a regular MySQL instance (all good):

```
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 21
Server version: 8.1.0 Homebrew

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> SELECT NULL AS number_nullable, NULL AS string_nullable, NULL AS date_nullable
    -> UNION ALL
    -> SELECT 123 AS number_nullable, 'abc' AS string_nullable, CURDATE() AS date_nullable;
+-----------------+-----------------+---------------+
| number_nullable | string_nullable | date_nullable |
+-----------------+-----------------+---------------+
|            NULL | NULL            | NULL          |
|             123 | abc             | 2023-10-31    |
+-----------------+-----------------+---------------+
2 rows in set (0.01 sec)
```

To the MySQL Clickhouse instance (also all good!):

```
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 23
Server version: 23.10.1.1795-ClickHouse

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> SELECT NULL AS number_nullable, NULL AS string_nullable, NULL AS date_nullable
    -> UNION ALL
    -> SELECT 123 AS number_nullable, 'abc' AS string_nullable, CURDATE() AS date_nullable;
+-----------------+-----------------+---------------+
| number_nullable | string_nullable | date_nullable |
+-----------------+-----------------+---------------+
|             123 | abc             | 2023-10-31    |
|            NULL | NULL            | NULL          |
+-----------------+-----------------+---------------+
2 rows in set (0.10 sec)
Read 2 rows, 2.00 B in 0.001288 sec., 1552 rows/sec., 1.52 KiB/sec.
```

**Describe what's wrong**

Tableau converts null in a nullable number or date field to 0 instead of showing it as `NULL`. Strings work as expected.

Is there anything we can do on the Clickhouse side to make Tableau understand the schema better? 

This might very well be a Tableau bug, but hopefully there is a way to fix it from the Clickhouse side.

**How to reproduce**

* Which ClickHouse server version to use: latest master
* Which interface to use: MySQL

**Expected behavior**

The MySQL Clickhouse connector should behave the same way as a regular MySQL instance in tableau when it comes to parsing `nullable` values

### PR DESCRIPTION
**Pull request by slvrtrn**
Resolves #56180

This was caused by the `->isNullAt` behavior with non-Nullable, which always returned `false` after we obtained a pointer to the nested column; now, it is processed correctly. 

* `MySQLUtils::getBaseColumn` was removed, as it caused more harm than good. 
* Added more tests to verify that the nullable flag is correctly set for a fetched column.
* Reference test results for the Java client were unified, as now both text/binary protocol tests produce the same results.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed incorrect nullable columns reporting in MySQL binary protocol, which caused certain 3rd party tools such as Tableau Online to display the default values instead of `NULL`.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 9e3ba550701f68295774b7992e6261cd94da2167 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56799/9e3ba550701f68295774b7992e6261cd94da2167/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6879771569">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56799/9e3ba550701f68295774b7992e6261cd94da2167/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Core/MySQL/PacketsProtocolBinary.cpp [hunk 1]
```diff
@@ -30,17 +25,21 @@ ResultSetRow::ResultSetRow(const Serializations & serializations_, const DataTyp
     : row_num(row_num_), columns(columns_), data_types(data_types_), serializations(serializations_)
 {
     payload_size = 1 + null_bitmap_size;
-    FormatSettings format_settings;
+    static FormatSettings format_settings;
```
Comment (rschu1ze) [inline src/Core/MySQL/PacketsProtocolBinary.cpp:28]: Please either remove `static` or add a comment that `static` may cause eye cancer :-)

**Comment by slvrtrn** [comment]
CC @larskarbo

File: src/Core/MySQL/PacketsProtocolText.cpp [hunk 2]
```diff
@@ -1,12 +1,13 @@
+#include <Columns/ColumnNullable.h>
```
Comment (rschu1ze) [inline src/Core/MySQL/PacketsProtocolText.cpp:1]: By convention, the corresponding header file of a source file (PacketsProtocolText.h) is included always first.

**Review by rschu1ze — APPROVED** [review]

**Pull request merged**